### PR TITLE
feat: sample push down

### PR DESCRIPTION
**Pull request by gaurav274**

### CODE REVIEW

File: eva/readers/opencv_reader.py [hunk 1]
```diff
@@ -44,12 +53,26 @@ def _read(self) -> Iterator[Dict]:
...
+                    yield {"id": frame_id, "data": frame}
+                    _, frame = video.read()
+                    frame_id += 1
+        else:
+            for begin, end in range_list:
+
+                # align begin with sampling rate
+                if begin % self._sampling_rate:
+                    begin += self._sampling_rate - (begin % self._sampling_rate)
+                print(begin, end + 1, self._sampling_rate)
```
Comment (xzdandy) [inline eva/readers/opencv_reader.py:71]: Remove print

File: eva/readers/opencv_reader.py [hunk 2]
```diff
@@ -44,12 +53,26 @@ def _read(self) -> Iterator[Dict]:
...
+                while frame is not None and frame_id <= end:
+                    yield {"id": frame_id, "data": frame}
+                    _, frame = video.read()
+                    frame_id += 1
+        else:
+            for begin, end in range_list:
+
+                # align begin with sampling rate
+                if begin % self._sampling_rate:
+                    begin += self._sampling_rate - (begin % self._sampling_rate)
```
Comment (xzdandy) [inline eva/readers/opencv_reader.py:70]: If sampling has higher precedence, should it happen before predicate? Is that due to the fact that we only push down range id predicates, so the order does not matter after we align the `begin`?

File: eva/readers/abstract_reader.py [hunk 3]
```diff
@@ -57,6 +57,7 @@ def read(self) -> Iterator[Batch]:
                 row_size = get_size(data)
             data_batch.append(data)
             if len(data_batch) * row_size >= self.batch_mem_size:
+                print("yielding: ", len(data_batch), self.batch_mem_size, row_size)
```
Comment (pchunduri6) [inline eva/readers/abstract_reader.py:60]: Remove print

File: eva/readers/opencv_reader.py (hunk 2 above)
Comment (gaurav274) [reply eva/readers/opencv_reader.py:70]: Yes. And I also removed the PushFilterBelowSample Rule to enforce it

**Review by xzdandy — APPROVED** [review]

**Pull request merged**